fili-core/src/main/java/com/yahoo/bard/webservice/data/config/metric/parser/MetricParser.java

```diff
@@ -0,0 +1,297 @@
+     * @throws ParsingException when a parsing error occurs
+     */
+    private int parseFuncArgs() throws ParsingException {
+        int argc = 0;
+
+        while (true) {
+            operatorStack.push(new Sentinel());
+            parseExpression();
+            argc += 1;
+            operatorStack.pop();
+
+            if (!peekType(LexemeType.COMMA)) {
+                break;
+            }
+            expect(LexemeType.COMMA);
+        }
+
+        return argc;
+    }
+
+    /**
+     * Get the next item on the top of the lexeme stack if it matches `type`, or throw an exception.
+     *
+     * @param type Lexeme type
+     * @return the item of the specified type
+     * @throws ParsingException when a parsing error occurs
+     */
+    private Lexeme expect(LexemeType type) throws ParsingException {
+        Lexeme current = items.remove();
+        if (!(current.getType() == type)) {
+            throw new ParsingException("Unexpected token type: found " + current.getType() + " but expected " + type);
+        }
+        return current;
+    }
+
+    /**
+     * Return true if the lexeme on the top of the stack equals the passed-in type.
+     *
+     * @param type lexeme type
+     * @return true if the top item matches, false otherwise
+     */
+    private boolean peekType(LexemeType type) {
+        return items.size() != 0 && (items.peek().getType() == type);
+    }
+
+    /**
+     * Return the next lexeme if it matches `type`, or null.
+     *
+     * @param type lexeme type
+     * @return item if the top item's type matches, null otherwise
+     * @throws ParsingException when a parsing error occurs
+     */
+    private Lexeme peekPop(LexemeType type) throws ParsingException {
+        if (peekType(type)) {
+            return expect(type);
+        }
+
+        return null;
+    }
+
+    /**
+     * Operator stack pop.
+     */
+    private void popOperator() {
+        if (operatorStack.peek().getPrecedence().greaterThan(Precedence.SENTINEL)) {
+            Operator operator = operatorStack.pop();
+            LinkedList<Operand> operands = new LinkedList<>();
+
+            for (int i = 0; i < operator.getNumOperands(); i++) {
+                operands.addFirst(operandStack.pop());
+            }
+
+            operandStack.push(operator.build(operands));
+        }
+    }
+
+    /**
+     * Operator stack push.
+     *
+     * @param op the operator to add to the stack
+     */
+    private void pushOperator(Operator op) {
+        while (true) {
+            if (operatorStack.peek().greaterThan(op)) {
+                popOperator();
+            } else {
+                break;
+            }
+        }
+        operatorStack.push(op);
+    }
+
+    /**
+     * Generate a globally unique new name.
+     * <p>
+     * Note: this assumes that we've chosen a sufficiently unique prefix.
+     *
+     * @return the new name
+     */
+    public synchronized static String getNewName() {
```

Comment:
This is my least favorite line of code in the whole thing, but I forgot to add a FIXME.